First (almost) complete version of script, use of Thing Model as an input

JKRhb opened the pull request:
Besides further developing the generation script, I changed to input files to a thing model (see #13) and instance specific information which are then merged for creating the C code. 

Using a thing model should provide the benefit to be able to reuse existing thing models (possibly even adapted from legacy self description standards) and adapt them to RIOT by providing forms and security information. The security definitions the thing and its affordances can pick from, however, should be provided by RIOT modules themselves and could be gradually extended by offering some kind of interface for inserting them into our script.

As in the current version of the script handlers are supposed to be included via adding them to the `forms`, and only `securityDefinitions` are handed over via script parameters, I removed the possibility to use multipe JSON files as input. This is, of course, up for discussion but while implementing the script I didn't see that much of a benefit. 

Regarding the SAUL integration, I would suggest that we use some kind of special key (something like `riot_os: saul_handler`) which can then be used to import an appropriate SAUL handler. So far, however, I have almost no experience regarding SAUL so I am curious what you think about this and the implementation as a whole :)

PS: At the moment, the example application does not compile because I changed some elements of `wot.h` to be closer to the specification and couldn't exactly figure out yet how to adjust all parts of `serialization.c` accordingly. This, however, should be easy to fix for you :)

JKRhb committed:
Add a thing model and instance information as example

JKRhb committed:
Add allowed operations for the thing level

JKRhb committed:
Refactor adding descriptions and titles

JKRhb committed:
Change datastructure for media type

JKRhb committed:
Finish first final version of script, change input to thing model and instance information

Citrullin commented:
> As in the current version of the script, handlers are supposed to be included via adding them to the `forms`, and only `securityDefinitions` are handed over via script parameters I removed the possibility to use multipe JSON files as input. This is, of course, up for discussion but while implementing the script I didn't see that much of a benefit.

Let's see, if it makes problems at some point. If so, we are still able to add it.

JKRhb committed:
Add default language tag

JKRhb committed:
Update wot_config.c

Citrullin commented on tests/wot_coap_generation/config/wot_td/instance.json:
```diff
@@ -0,0 +1,29 @@
+{
+    "thing": {
```
Okay, it gets the title and description from the model, if I get this right. But don't we need some specifier to identify the device itself? The `id` field is probably good for that. Otherwise I am not able to identify the device.

JKRhb replied:
Hmm, yeah, you are right. The `id` field, however, needs to be an `anyURI` (see [here](https://www.w3.org/TR/wot-thing-description/#thing)) – maybe we could, alternatively, just insert the `base` URI a second time during the TD creation process at runtime?

JKRhb committed:
Merge branch 'wot_gcoap' into wot_gcoap_thing_model

JKRhb committed:
Fix generation of default_language string

JKRhb committed:
Update wot_config.c

JKRhb committed:
Fix generation of default_language string

JKRhb committed:
Remove walrus operator from script for better compatibility

JKRhb committed:
Make first attempt of integrating generation script into build process

@Citrullin This is probably not the right way to do it (as it also does not
completely work right now) but it is a start, I guess :)

Citrullin committed:
Fix issues in wot_config.c

Citrullin committed:
Add fixme

Citrullin committed:
Some serilization fixes

Citrullin committed:
Fix merge conflicts

Citrullin commented:
@JKRhb Apparently Signal has some issues due to tons of new users. I pushed some commits into your branch. You are now able to serialize everything as before.

JKRhb committed:
Add possibility to add thing models from build process

JKRhb committed:
Remove generation script from the tests directory

JKRhb commented:
> 
> 
> @JKRhb Apparently Signal has some issues due to tons of new users. I pushed some commits into your branch. You are now able to serialize everything as before.

Awesome, thank you for the fixes! :)

In the meantime, I managed to implement a first version of an integration of the script into the build process (see https://github.com/Citrullin/RIOT/pull/14/commits/ffe497fc0af7551b1ba990a6c1142e7c1a0bd02f). It is actually already possible to integrate multiple thing models as command line arguments so I think by using this method we could implement our modular approach very elegantly, e. g. by providing thing models for each type of sensor supported by SAUL. 

There are, however, still some issues with the execution order of the script which is probably caused by placing it into the wrong makefile. I hope this is easy to fix, though.

JKRhb committed:
Pass used modules to wot generation script

JKRhb committed:
Move generation Makefile to "makefiles" folder

JKRhb committed:
Add thing_instance_information to makefile and script

JKRhb committed:
Move definition of json_files to app makefile

JKRhb reviewed (commented):
@Citrullin I think I found an elegant solution for integrating thing models and instance info into the script. At the moment, however, it is not executed automatically but only by using the  command `make wot-config`. But maybe this is actually more desirably then automatic execution?

Citrullin committed:
Fix included null pointer

JKRhb committed:
Remove resolved TODOs from generaion script

JKRhb committed:
Merge branch 'wot_gcoap_thing_model' of github.com:Citrullin/RIOT into wot_gcoap_thing_model

JKRhb committed:
Use "no security" as default scheme when no security is defined

JKRhb committed:
Update wot_config.c

JKRhb committed:
Implement proper timezone offset in generation script

JKRhb committed:
Remove debug print statements

JKRhb committed:
Add missing "href" field to link struct

JKRhb committed:
Add functionality to include multiple thing models and merge them

JKRhb committed:
Update example

Citrullin committed:
Refactor makefiles

Citrullin committed:
Refactor makefiles

JKRhb committed:
Export WOT variables

JKRhb committed:
Use python3 for build process

JKRhb committed:
Update python script to include app directory

Citrullin committed:
Fix some issues

JKRhb committed:
Fix error regarding the generation of type fields

JKRhb committed:
Add new makefile rule for generation of wot config

"all" depends now on this rule

JKRhb committed:
Replace SRC with GENSRC

Citrullin committed:
Some fixes

Citrullin committed:
Fix config

JKRhb committed:
Merge pull request #22 from Citrullin/wot_makefile

Merge Makefile branch into thing_model branch

JKRhb committed:
Remove unused code

JKRhb committed:
Add proper check for required affordances of thing models

JKRhb committed:
Add (hopefully) correct sorting of CoAP resources

JKRhb committed:
Add TODO for inserting link parameters

JKRhb committed:
Revert "Add (hopefully) correct sorting of CoAP resources"

This reverts commit 4488075b61c49f719e32133b15819ac2516012df.

JKRhb committed:
Remove unneeded functions from generation script

JKRhb committed:
Allow for hrefs with hypens

JKRhb committed:
Allow for partial implementation of thing models

JKRhb committed:
Use namespace "riot_os" for RIOT specific settings in instance JSON

JKRhb committed:
Prevent @context duplicates

JKRhb committed:
Let build process on JSON files, remove thing model merging

JKRhb committed:
Retrieve default @language from @context (if it is set)

JKRhb committed:
Fix rule order in Makefile.gensrc

JKRhb committed:
Remove trailing whitespace

Citrullin committed:
Refactoring wot_forms in coap

Citrullin committed:
Fix security in wot definition

Citrullin committed:
Fix missing character for type

JKRhb committed:
Update generation script

JKRhb committed:
Remove obsolete add_hef function from generation script

JKRhb committed:
Update example JSON

JKRhb committed:
Add determination of default CoAP methods

JKRhb committed:
Fix issue with missing key of affordances

JKRhb committed:
Remove extension struct

Citrullin committed:
Merge pull request #27 from Citrullin/wot_form_refactoring

Refactoring wot_forms in coap

JKRhb committed:
Merge branch 'wot_gcoap_thing_model' into wot_security_fixes

JKRhb committed:
Fix security definitions in generation script

JKRhb committed:
Merge pull request #28 from Citrullin/wot_security_fixes

Wot security fixes

JKRhb committed:
Merge branch 'wot_gcoap_thing_model' into fix_type

JKRhb committed:
Merge branch 'wot_gcoap_thing_model' into fix_type

JKRhb committed:
Refactor copying of "forms" affordance field

JKRhb committed:
Copy security from the instance information

JKRhb committed:
Autoformat generation script

JKRhb committed:
Merge pull request #29 from Citrullin/fix_type

Fix type

JKRhb committed:
Fix link parameters of TD resource

See https://github.com/w3c/wot-discovery/issues/120

JKRhb committed:
Add brackets around IPv6 base address

JKRhb committed:
Merge branch 'master' into wot_gcoap_thing_model

JKRhb committed:
Merge branch 'wot_gcoap' into wot_gcoap_thing_model

JKRhb committed:
Remove "zero_struct" CStruct option from python script

JKRhb committed:
Fix typos in python script

JKRhb committed:
Fix python style issues

JKRhb committed:
Fix whitespace error in coap_config/Makefile.gensrc

JKRhb committed:
Remove trailing whitespace

JKRhb committed:
Fix more typos

JKRhb committed:
Add C++ compatible headers

JKRhb committed:
Fix header guard formats

JKRhb committed:
fixup! Fix header guard formats

JKRhb committed:
Fix variable scopes

JKRhb committed:
Fix whitespace and newline issues

JKRhb committed:
Fix header guards

JKRhb committed:
Fix unneeded variable reassignment

JKRhb committed:
Remove APPLICATION from test Makefiles

JKRhb committed:
Add ARRAY_SIZE to old wot_coap test
